### PR TITLE
tetragon: Add bpf program loader tests

### PR DESCRIPTION
**Pull request by olsajiri**
Adding CheckSensorLoad function that takes array of programs
and maps and checks they are loaded and share specified maps.

Plus related changes to allow that.
    
Signed-off-by: Jiri Olsa <jolsa@kernel.org>

### CODE REVIEW

File: pkg/sensors/program/coll.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+//     copyLoadedCollection(lc, coll)
+//
+//   - collection is removed and only 'used' programs and maps stay loaded
+//     so we filter out unused programs/maps with
+//
+//     load.LC = filterLoadedCollection(lc)
+//
+// This was we have only programs/maps realted to our test and with
+// original long names.
+//
```
Comment (kkourt) [inline pkg/sensors/program/coll.go:40]: Thanks for writing this down!

One thing that got me thinking is the following. You say that:
> // We can't just iterate all load programs and maps for 2 reasons:
> //   - we could race with others loading bpf programs

Couldn't we race with other programs between copyLoadedCollection() and filterLoadedCollection()? If we only use this for testing, the race might be acceptable but I'm curious.

File: pkg/sensors/program/coll.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+//   - copy all programs/maps from ebpf collection
+//
+//     copyLoadedCollection(lc, coll)
+//
+//   - collection is removed and only 'used' programs and maps stay loaded
+//     so we filter out unused programs/maps with
+//
+//     load.LC = filterLoadedCollection(lc)
+//
+// This was we have only programs/maps realted to our test and with
```
Comment (kkourt) [inline pkg/sensors/program/coll.go:38]: Not sure I fully understand the above. How about:
> Now load.LC contains programs/maps that are related to our test, are actually loaded, and have their original long names.

File: pkg/sensors/program/coll.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	for name, lp := range lc.Programs {
+		logger.GetLogger().Infof("%d: %s - %v", lp.ID, name, lp.MapIDs)
+	}
+	logger.GetLogger().Infof("Maps (%s):", str)
+	for name, lm := range lc.Maps {
+		logger.GetLogger().Infof("%d: %s", lm.ID, name)
+	}
+}
+
+func copyLoadedCollection(lc *LoadedCollection, coll *ebpf.Collection) error {
```
Comment (kkourt) [inline pkg/sensors/program/coll.go:81]: how about writing this as:
> copyLoadedCollection(coll *ebpf.Collection) (*LoadedCollection, error) 

I think it makes the interface clearer (?)

File: pkg/sensors/program/coll.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		lp.MapIDs = mapIDs
+		lp.Type = p.Type()
+		lc.Programs[name] = lp
+	}
+	return nil
+}
+
+// Gets all the programs/maps and removes any non existent ones
+// from passed collection
+func filterLoadedCollection(lc *LoadedCollection) *LoadedCollection {
```
Comment (kkourt) [inline pkg/sensors/program/coll.go:122]: nit: how about writing this as:
> func (lc *LoadedCollection) filter() *LoadedCollection {
> ...
> }

We coud even do
> func (lc *LoadedCollection) filter() {
> ...
> }

And do the changes in-place.

File: pkg/observer/observer.go [hunk 5]
```diff
@@ -252,6 +252,8 @@ type Observer struct {
 
 	/* YAML Configuration File */
 	configFile string
+
+	Sensors []*sensors.Sensor
```
Comment (kkourt) [inline pkg/observer/observer.go:256]: It fees a bit strange adding this here. The way that the observer is supposed to manage sensor is via the "sensor manager" which is a goroutine for managing sensors.

See: https://github.com/cilium/tetragon/blob/bb6f985d32a3a5777d9c922961c77ac5d0bdbfb9/pkg/sensors/sync.go#L24-L30

From what I can tell, this is for testing only, so maybe we can keep the list elsewhere if we need to?

**Review by kkourt — COMMENTED** [review]
Thanks! LGTM! This is definitely needed.

I have some comments. None of them is a blocker, but might be worthwhile discussing them before merging.

File: pkg/sensors/program/coll.go (hunk 1 above)
Comment (olsajiri) [reply pkg/sensors/program/coll.go:40]: as we discussed later the filtering is based on program/map ID, so it won't race

File: pkg/sensors/program/coll.go (hunk 2 above)
Comment (olsajiri) [reply pkg/sensors/program/coll.go:38]: LC contains arrays of program and maps that are related to the object we loaded plus we filtered out unused program and maps that we unloaded.. all this with original long names for both programs and maps

File: pkg/sensors/program/coll.go (hunk 3 above)
Comment (olsajiri) [reply pkg/sensors/program/coll.go:81]: yes, I will put the LoadedCollection as return value, thanks

File: pkg/observer/observer.go (hunk 5 above)
Comment (olsajiri) [reply pkg/observer/observer.go:256]: as discussed later I'll change this to have sensors are return value

**Review by kkourt — APPROVED** [review]

**Pull request merged**